apps/prairielearn/src/ee/lib/aiQuestionGeneration.ts

```diff
@@ -3,72 +3,129 @@ import { OpenAI } from 'openai';
 
 import { createEmbedding, openAiUserFromAuthn, vectorToString } from './contextEmbeddings.js';
 
 const sql = loadSqlEquiv(import.meta.url);
 
 /**
- * Generates the HTML and Python code for a new question using an LLM
+ * Generates the common preamble with general PrairieLearn information for the LLM
  *
- * @param client the OpenAI client to use
- * @param courseId the ID of the current course
- * @param authnUserId the authenticated user's ID
- * @param prompt the prompt for how to generate a question
- * @returns a server job ID for the generation task
+ * @param context Relevant example documents, formatted into one string.
+ * @returns A string, the prompt preamble.
+ */
+function promptPreamble(context: string): string {
+  return `# Introduction
+        
+        You are an assistant that helps instructors write questions for PrairieLearn.
+        
+        A question has a \`question.html\` file that can contain standard HTML, CSS, and JavaScript. It also includes PrairieLearn elements like \`<pl-multiple-choice>\` and \`<pl-number-input>\`.
+        
+        A question may also have a \`server.py\` file that can randomly generate unique parameters and answers, and which can also assign grades to student submissions. \`server.py\` may be omitted if it's not necessary.
+        
+        ## Generating random parameters
+        
+        \`server.py\` may define a \`generate\` function. \`generate\` has a single parameter \`data\` which can be modified by reference. It has the following properties:
+        
+        - \`params\`: A dictionary. Random parameters, choices, etc. can be written here for later retrieval.
+        
+        ## Using random parameters
+        
+        Parameters can be read in \`question.html\` with Mustache syntax. For instance, if \`server.py\` contains \`data["params"]["answer"]\`, it can be read with \`{{ params.answer }}\` in \`question.html\`.
+        
+        # Context
+        
+        Here is some context that may help you respond to the user. This context may include example questions, documentation, or other information that may be helpful.
+        
+        ${context}
+        
+        `;
+}
+
+/**
+ * Builds the context string, consisting of relevant documents.
+ *
+ * @param client The OpenAI client to use.
+ * @param prompt The user's question generation prompt.
+ * @param authnUserId The user's authenticated user ID.
+ * @returns A string of all relevant context documents.
+ */
+async function makeContext(client: OpenAI, prompt: string, authnUserId: string): Promise<string> {
+  const embedding = await createEmbedding(client, prompt, openAiUserFromAuthn(authnUserId));
+
+  const docs = await queryRows(
+    sql.select_nearby_documents,
+    { embedding: vectorToString(embedding), limit: 5 },
+    QuestionGenerationContextEmbeddingSchema,
+  );
+
+  const contextDocs = docs.map((doc) => doc.doc_text);
+  const context = contextDocs.join('\n\n');
+  return context;
```

Comment:
Nit, a little more concise:

```suggestion
  return docs.map((doc) => doc.doc_text).join('\n\n');
```